packages/apollo-server-core/src/plugin/schemaReporting/index.ts

```diff
@@ -0,0 +1,130 @@
+  /**
+   * The URL to use for reporting schemas. Primarily for testing and internal
+   * Apollo use.
+   */
+  endpointUrl?: string;
+}
+
+export function ApolloServerPluginSchemaReporting(
+  {
+    initialDelayMaxMs,
+    overrideReportedSchema,
+    endpointUrl,
+  }: ApolloServerPluginSchemaReportingOptions = Object.create(null),
+): ApolloServerPlugin {
+  const bootId = uuidv4();
+
+  return {
+    __internal_plugin_id__() {
+      return 'SchemaReporting';
+    },
+    async serverWillStart({ apollo, schema, logger }) {
+      if (!apollo.key) {
+        throw Error(
+          'To use ApolloServerPluginSchemaReporting, you must provide an Apollo API ' +
+            'key, via the $APOLLO_KEY environment variable or via `new ApolloClient({apollo: {key})`',
+        );
+      }
+      if (!apollo.graphId) {
+        throw Error(
+          "To use ApolloServerPluginSchemaReporting, you must provide your graph's ID, " +
+            "either by using an API key starting with 'service:',  or by providing it explicitly via " +
+            'the $APOLLO_GRAPH_ID environment variable or via `new ApolloClient({apollo: {graphId}})`',
+        );
+      }
+
+      const executableSchema = overrideReportedSchema ?? printSchema(schema);
+      const executableSchemaId = computeExecutableSchemaId(executableSchema);
+
+      logger.info('Starting schema reporter...');
+      if (overrideReportedSchema !== undefined) {
+        logger.info('Schema to report has been overridden');
+      }
+      if (initialDelayMaxMs !== undefined) {
+        logger.info(
+          `Schema reporting max initial delay override: ${initialDelayMaxMs} ms`,
+        );
+      }
+      if (endpointUrl !== undefined) {
+        logger.info(`Schema reporting URL override: ${endpointUrl}`);
+      }
+
+      const serverInfo = {
+        bootId,
+        graphVariant: apollo.graphVariant,
+        // The infra environment in which this edge server is running, e.g. localhost, Kubernetes
+        // Length must be <= 256 characters.
+        platform: process.env.APOLLO_SERVER_PLATFORM || 'local',
+        runtimeVersion: `node ${process.version}`,
+        executableSchemaId: executableSchemaId,
+        // An identifier used to distinguish the version of the server code such as git or docker sha.
+        // Length must be <= 256 charecters
+        userVersion: process.env.APOLLO_SERVER_USER_VERSION,
+        // "An identifier for the server instance. Length must be <= 256 characters.
+        serverId:
+          process.env.APOLLO_SERVER_ID || process.env.HOSTNAME || os.hostname(),
+        libraryVersion: `apollo-server-core@${
+          require('../../../package.json').version
+        }`,
+      };
+
+      logger.info(
+        `Schema reporting EdgeServerInfo: ${JSON.stringify(serverInfo)}`,
+      );
+
+      const schemaReporter = new SchemaReporter({
+        serverInfo,
+        schemaSdl: executableSchema,
+        apiKey: apollo.key!,
```

Comment:
Seems like this shouldn't be necessary, we assert `apollo.key`'s existence above.

Reply:
It is necessary (presumably TypeScript isn't positive that apollo.key isn't from some sort of getter?) but I can fix it in another way by saving to a const.